src/api-service/__app__/onefuzzlib/agent_events.py

```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env python
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import logging
+from typing import Optional, Union, cast
+from uuid import UUID
+
+from onefuzztypes.enums import (
+    ErrorCode,
+    NodeState,
+    NodeTaskState,
+    TaskDebugFlag,
+    TaskState,
+)
+from onefuzztypes.models import (
+    Error,
+    NodeDoneEventData,
+    NodeSettingUpEventData,
+    NodeStateUpdate,
+    WorkerDoneEvent,
+    WorkerEvent,
+    WorkerRunningEvent,
+)
+
+from ..onefuzzlib.pools import Node, NodeTasks
+from ..onefuzzlib.task_event import TaskEvent
+from ..onefuzzlib.tasks.main import Task
+
+
+def get_node(machine_id: UUID) -> Union[Node, Error]:
+    node = Node.get_by_machine_id(machine_id)
+    if not node:
+        return Error(code=ErrorCode.INVALID_NODE, errors=["unable to find node"])
+    return node
+
+
+def on_state_update(
+    machine_id: UUID,
+    state_update: NodeStateUpdate,
+) -> Optional[Error]:
+    state = state_update.state
+    node = get_node(machine_id)
+    if isinstance(node, Error):
+        return node
+
+    if state == NodeState.free:
+        if node.reimage_requested or node.delete_requested:
+            logging.info("stopping free node with reset flags: %s", node.machine_id)
+            node.stop()
+            return None
+
+        if node.could_shrink_scaleset():
+            logging.info("stopping free node to resize scaleset: %s", node.machine_id)
+            node.set_halt()
+            return None
+
+    if state == NodeState.init:
+        if node.delete_requested:
+            logging.info("stopping node (init and delete_requested): %s", machine_id)
+            node.stop()
+            return None
+        node.reimage_requested = False
+        node.state = state
+        node.save()
+        return None
+
+    logging.info("node state update: %s from:%s to:%s", machine_id, node.state, state)
+    node.state = state
+    node.save()
+
+    if state == NodeState.setting_up:
+        # Model-validated.
+        #
+        # This field will be required in the future.
+        # For now, it is optional for back compat.
+        setting_up_data = cast(
+            Optional[NodeSettingUpEventData],
+            state_update.data,
+        )
+
+        if setting_up_data:
+            if not setting_up_data.tasks:
+                raise Exception("setup without tasks.  machine_id: %s", machine_id)
+
+            for task_id in setting_up_data.tasks:
+                task = Task.get_by_task_id(task_id)
+                if isinstance(task, Error):
```

Comment:
If we hit this error case, what should happen to the node? Right now, we would have already committed the state update, and we don't attempt to explicitly fail it or roll back.

Reply:
We moved to node state management to solely be at the purview of the supervisor.  Even with the exception in the weird case, the node state still is set to whatever the node declared.

Reply:
Okay, I buy that. That means the user-facing node state is what the supervisor last claimed it to be, whether or not that is in some sense inaccurate.

If we accept that, then my only complaint would be that if a node is running 2 tasks, and we can't `get_by_id()` the 1st one, but we _could_ for the 2nd one, we'll fail to update the 2nd task because we return early, even though something has only gone wrong with the 1st `task`. But that's not unique to this PR, and that case is actually currently unreachable, so shouldn't be addressed now.